libsolidity/ast/Types.cpp

```diff
@@ -3466,15 +3466,19 @@ Type const* FunctionType::mobileType() const
 	if (valueSet() || gasSet() || saltSet() || hasBoundFirstArgument())
 		return nullptr;
 
-	// return function without parameter names
+	// Special function types do not get a mobile type, such that they cannot be used in complex expressions.
+	if (m_kind != FunctionType::Kind::Internal && m_kind != FunctionType::Kind::External && m_kind != FunctionType::Kind::DelegateCall)
+		return nullptr;
```

Comment:
I found something that could be potentially broken by this PR.

```solidity
contract C {
    function f() internal {}
    function g() internal {}

    function test() public returns (bool) {
        function () internal ptr = C.f;
        return ptr == C.g;
    }
}
```

Here we're comparing `ptr` (`FunctionType::Kind::Internal`) with an internal function accessed via contract name which, I assume should be `FunctionType::Kind::Declaration`. It's allowed on 0.8.20, though I haven't checked if it the comparison gives correct results (I assume so).

After this PR `C.g` will get no mobile type and they won't be comparable anymore. We should verify that. If that's the case, the PR is breaking for function type comparisons.

Reply:
I added this as a test case and it looks like it is not breaking. From what I found, `C.f` and `C.g` both have `FunctionType::Kind::Internal`.

Reply:
Uh, I guess I spoke too soon. But something weird is happening, the obtained result is always the opposite of the expected.
Upon updating the expectation and re-running the test, the obtained result changes again...

Reply:
This means it's broken. What you're seeing is probably legacy running first, giving you one result. Then you accept the update, IR version runs and gives you the opposite result. So they give you different results for the comparison.

Reply:
Are you sure ``C.g`` has a declaration function type here? I'd have assumed all types involved here were internal function types.
This may be subject to https://github.com/ethereum/solidity/issues/13969, but I'd assume that this won't be affected by the PR.

Reply:
> Are you sure `C.g` has a declaration function type here? I'd have assumed all types involved here were internal function types.

I am not, and you're right, this is an internal type. I guess we can consider this one resolved.

Well, I just don't have a good grasp of which functions are of `Declaration` kind. Are these just internal functions of foreign contracts or is there something else?

I find this aspect of the type system really counterintuitive and it's hard to check that in code - `Declaration` is the default kind for `FunctionType()` so you can't easily find places where it's applied. Looking at this https://github.com/ethereum/solidity/blob/96bb39d1b4569b8eb3f00273cce7741d9f55a662/libsolidity/ast/AST.cpp#L474-L485 I assumed that any function accessed via contract name would be a `Declaration` but apparently this is not the case.

> This may be subject to https://github.com/ethereum/solidity/issues/13969, but I'd assume that this won't be affected by this PR at all.

ok, it's likely that this is exactly what was happening here.

Reply:
> Well, I just don't have a good grasp of which functions are of `Declaration` kind. Are these just internal functions of foreign contracts or is there something else?

For future reference, looks like @ekpyron actually explained that earlier: https://github.com/ethereum/solidity/pull/14365/files#r1243712169

That's still not a comprehensive answer though and there can be other cases I guess?